contracts/modules/v2/StakingPool.sol

```diff
@@ -173,145 +174,238 @@ contract StakingPool is IStakingPool, ERC721 {
     activeStake = _activeStake + amount;
     stakeSharesSupply = _stakeSharesSupply + newStakeShares;
     rewardsSharesSupply = _rewardsSharesSupply + newRewardsShares;
+
+    // update group and staker data
+    groups[groupId] = group;
+    positions[tokenId][groupId] = position;
+    pendingRewards[tokenId] = pendingReward;
+
+    /* end storage writes */
   }
 
-  function withdraw(WithdrawParams[] memory params) external {
-    // 1. check nft ownership / allowance
-    // 2. loop through each group:
-    // 2.1. check group expiration if the deposit flag is set
-    // 2.2. calculate the reward amount if the reward flag is set
-    // 3. sum up nxm amount of each group
-    // 4. transfer nxm to staker
+  function calculateRewardSharesAmount(
+    uint stakeSharesAmount,
+    uint groupId
+  ) internal view returns (uint) {
+
+    uint lockDuration = (groupId + 1) * GROUP_SIZE - block.timestamp;
+    uint maxLockDuration = GROUP_SIZE * 8;
+
+    // TODO: determine extra rewards formula
+    return
+      stakeSharesAmount
+      * REWARDS_SHARES_RATIO
+      * lockDuration
+      / REWARDS_SHARES_DENOMINATOR
+      / maxLockDuration;
+  }
+
+  function withdraw(WithdrawParams[] calldata params) external {
+
+    updateGroups();
+
+    uint _accNxmPerRewardsShare = accNxmPerRewardsShare;
+    uint _firstActiveGroupId = block.timestamp / GROUP_SIZE;
+
+    for (uint i = 0; i < params.length; i++) {
+
+      uint stakeToWithdraw;
+      uint rewardToWithdraw;
+
+      uint tokenId = params[i].tokenId;
+      uint groupCount = params[i].groupIds.length;
+
+      for (uint j = 0; j < groupCount; j++) {
+
+        uint groupId = params[i].groupIds[j];
+        Position memory position = positions[tokenId][groupId];
+
+        // can withdraw stake only if the group is expired
+        if (params[i].withdrawStake && groupId < _firstActiveGroupId) {
+
+          // calculate the amount of nxm for this position
+          uint stake = expiredGroups[groupId].stakeAmountAtExpiry;
+          uint stakeShareSupply = expiredGroups[groupId].stakeShareSupplyAtExpiry;
+          stakeToWithdraw += stake * position.stakeShares / stakeShareSupply;
+
+          // mark as withdrawn
+          position.stakeShares = 0;
+        }
+
+        if (params[i].withdrawRewards) {
+
+          // calculate reward since checkpoint
+          uint newRewardPerShare = _accNxmPerRewardsShare - position.lastAccNxmPerRewardShare;
+          rewardToWithdraw += newRewardPerShare * position.rewardsShares;
+
+          // save checkpoint
+          position.lastAccNxmPerRewardShare = _accNxmPerRewardsShare;
+        }
+
+        positions[tokenId][groupId] = position;
+      }
+
+      uint withdrawable = stakeToWithdraw + rewardToWithdraw;
+
+      if (params[i].withdrawRewards) {
+        withdrawable += pendingRewards[tokenId];
+        pendingRewards[tokenId] = 0;
+      }
+
+      // TODO: use TC instead
+      nxm.transfer(ownerOf(tokenId), withdrawable);
```

Comment:
Where are we minting the NXM tokens designated for rewards?

Reply:
I've left a bunch of todos in places where I'm touching/transfering nxm. All nxm transfer and minting ops will be handled by TokenController. When a cover is bought we'll mint the rewards so the token supply will correctly reflect the nxm that we'll distribute in the future.

Reply:
@roxdanila @danoctavian I don't think minting NXM is currently handled by Cover.sol the way @shark0der described it. Should we open a ticket for this?

Reply:
Issue created, as that's currently missing in the implementation: https://nexus-mutual.atlassian.net/jira/software/projects/NM/boards/1/backlog?selectedIssue=NM-574